firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +58,115 @@
+      super.onStop();
+      callbacks.run();
+    }
+
+    final CallbackList callbacks = new CallbackList();
+  }
+
+  @SuppressWarnings("deprecation")
+  public static class StopListenerFragment extends android.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      callbacks.run();
+    }
+
+    final CallbackList callbacks = new CallbackList();
+  }
+
+  @Nullable
+  private static <T> T castFragment(Class<T> fragmentClass, @Nullable Object fragment, String tag) {
+    try {
+      return fragmentClass.cast(fragment);
+    } catch (ClassCastException e) {
+      throw new IllegalStateException(
+          "Fragment with tag '"
+              + tag
+              + "' is a "
+              + fragment.getClass().getName()
+              + " but should be a "
+              + fragmentClass.getName());
+    }
+  }
+
+  private static final String SUPPORT_FRAGMENT_TAG = "FirestoreOnStopObserverSupportFragment";
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  /**
+   * Implementation for non-FragmentActivity Activity's. Unfortunatly, all Fragment related
+   * classes/methods with nonFragmentActivityActivity's are deprecated, implying that almost
+   * everything in this function is deprecated.
+   */
+  @SuppressWarnings("deprecation")
+  private void onActivityStopCallOnce(Activity activity, Runnable callback) {
+    hardAssert(
+        !(activity instanceof FragmentActivity),
+        "onActivityStopCallOnce must be called with a *non*-FragmentActivity Activity.");
+
+    activity.runOnUiThread(
+        () -> {
+          StopListenerFragment fragment =
+              castFragment(
+                  StopListenerFragment.class,
+                  activity.getFragmentManager().findFragmentByTag(FRAGMENT_TAG),
+                  FRAGMENT_TAG);
+
+          if (fragment == null || fragment.isRemoving()) {
+            fragment = new StopListenerFragment();
+            activity
+                .getFragmentManager()
+                .beginTransaction()
+                .add(fragment, FRAGMENT_TAG)
+                .commitNowAllowingStateLoss();
+          }
+
+          fragment.callbacks.add(callback);
+        });
+  }
+
+  private void onFragmentActivityStopCallOnce(FragmentActivity activity, Runnable callback) {
+    activity.runOnUiThread(
+        () -> {
+          StopListenerSupportFragment fragment =
+              castFragment(
+                  StopListenerSupportFragment.class,
+                  activity.getSupportFragmentManager().findFragmentByTag(SUPPORT_FRAGMENT_TAG),
+                  SUPPORT_FRAGMENT_TAG);
+
+          if (fragment == null || fragment.isRemoving()) {
+            fragment = new StopListenerSupportFragment();
+            activity
+                .getSupportFragmentManager()
+                .beginTransaction()
+                .add(fragment, SUPPORT_FRAGMENT_TAG)
+                .commitNowAllowingStateLoss();
```

Comment:
Turns out that commitNow* is api 24, so we can't use it. :(

We could use https://developer.android.com/reference/android/app/FragmentManager#executePendingTransactions(), but that will execute *all* pending transactions (incl completion callbacks), and could potentially be lengthy if the user has created a lot of fragment transactions.

That won't block the caller (unless the caller is already on the UI thread): activity.runOnUiThread() just schedules the block to be run rather than waiting for it to return.